File: apps/ios/Sources/Litter/Models/ConversationStatistics.swift
import Foundation

struct ConversationStatistics {
    var totalMessages: Int = 0
    var userMessageCount: Int = 0
    var assistantMessageCount: Int = 0
    var turnCount: Int = 0
    var commandsExecuted: Int = 0
    var commandsSucceeded: Int = 0
    var commandsFailed: Int = 0
    var filesChanged: Int = 0
    var totalCommandDurationMs: Int64 = 0
    var mcpToolCallCount: Int = 0
    var webSearchCount: Int = 0

    static func compute(from items: [HydratedConversationItem]) -> ConversationStatistics {
        var stats = ConversationStatistics()
        var seenTurnIds = Set<String>()

        for item in items {
            switch item.content {
            case .user:
                stats.userMessageCount += 1
                stats.totalMessages += 1
            case .assistant:
                stats.assistantMessageCount += 1
                stats.totalMessages += 1
            case .commandExecution(let data):
                stats.commandsExecuted += 1
                switch data.status {
                case .completed:
                    stats.commandsSucceeded += 1
                case .failed:
                    stats.commandsFailed += 1
                default:
                    break
                }
                if let duration = data.durationMs {
                    stats.totalCommandDurationMs += duration
                }
            case .fileChange:
                stats.filesChanged += 1
            case .mcpToolCall:
                stats.mcpToolCallCount += 1
            case .webSearch:
                stats.webSearchCount += 1
            default:
                break
            }

            if let turnId = item.sourceTurnId {
                seenTurnIds.insert(turnId)
            }
        }

        stats.turnCount = seenTurnIds.count

        return stats
    }
}

struct ServerUsageData {
    struct TokenEntry: Identifiable {
        let id = UUID()
        let threadTitle: String
        let tokens: UInt64
    }

    struct ActivityEntry: Identifiable {
        let id = UUID()
        let date: Date
        let turnCount: Int
    }

    struct ModelEntry: Identifiable {
        let id = UUID()
        let model: String
        let threadCount: Int
    }

    var tokensByThread: [TokenEntry] = []
    var activityByDay: [ActivityEntry] = []
    var modelUsage: [ModelEntry] = []
    var rateLimits: RateLimitSnapshot?

    static func compute(from threads: [AppThreadSnapshot], server: AppServerSnapshot) -> ServerUsageData {
        var data = ServerUsageData()
        data.rateLimits = server.rateLimits

        // Token usage by thread
        data.tokensByThread = threads.compactMap { thread in
            guard let tokens = thread.contextTokensUsed, tokens > 0 else { return nil }
            let title = thread.info.title ?? "Untitled"
            return TokenEntry(threadTitle: title, tokens: tokens)
        }.sorted { $0.tokens > $1.tokens }

        // Activity by day — aggregate thread creation/update timestamps
        var dayBuckets: [String: Int] = [:]
        let dayFormatter = DateFormatter()
        dayFormatter.dateFormat = "yyyy-MM-dd"

        for thread in threads {
            // Use updatedAt preferring over createdAt
            let ts = thread.info.updatedAt ?? thread.info.createdAt
            guard let ts else { continue }
            let date = Date(timeIntervalSince1970: TimeInterval(ts))
            let key = dayFormatter.string(from: date)
            dayBuckets[key, default: 0] += 1
        }

        data.activityByDay = dayBuckets.compactMap { key, count in
            guard let date = dayFormatter.date(from: key) else { return nil }
            return ActivityEntry(date: date, turnCount: count)
        }.sorted { $0.date < $1.date }

        // Model usage breakdown
        var modelCounts: [String: Int] = [:]
        for thread in threads {
            let model = thread.model ?? thread.info.model ?? "unknown"
            modelCounts[model, default: 0] += 1
        }

        data.modelUsage = modelCounts.map { model, count in
            ModelEntry(model: model, threadCount: count)
        }.sorted { $0.threadCount > $1.threadCount }

        return data
    }
}

File: apps/android/app/src/main/java/com/litter/android/ui/conversation/ConversationStatistics.kt
package com.litter.android.ui.conversation

import uniffi.codex_mobile_client.AppOperationStatus
import uniffi.codex_mobile_client.AppServerSnapshot
import uniffi.codex_mobile_client.AppThreadSnapshot
import uniffi.codex_mobile_client.HydratedConversationItem
import uniffi.codex_mobile_client.HydratedConversationItemContent
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

data class ConversationStatistics(
    val totalMessages: Int = 0,
    val userMessageCount: Int = 0,
    val assistantMessageCount: Int = 0,
    val turnCount: Int = 0,
    val commandsExecuted: Int = 0,
    val commandsSucceeded: Int = 0,
    val commandsFailed: Int = 0,
    val filesChanged: Int = 0,
    val totalCommandDurationMs: Long = 0,
    val mcpToolCallCount: Int = 0,
    val webSearchCount: Int = 0,
) {
    companion object {
        fun compute(items: List<HydratedConversationItem>): ConversationStatistics {
            var userMessages = 0
            var assistantMessages = 0
            var turns = 0
            var cmdsExecuted = 0
            var cmdsSucceeded = 0
            var cmdsFailed = 0
            var files = 0
            var cmdDurationMs = 0L
            var mcpCalls = 0
            var webSearches = 0

            var lastWasUser = false
            for (item in items) {
                when (val content = item.content) {
                    is HydratedConversationItemContent.User -> {
                        userMessages++
                        if (!lastWasUser) turns++
                        lastWasUser = true
                    }
                    is HydratedConversationItemContent.Assistant -> {
                        assistantMessages++
                        lastWasUser = false
                    }
                    is HydratedConversationItemContent.CommandExecution -> {
                        cmdsExecuted++
                        val status = content.v1.status
                        if (status == AppOperationStatus.COMPLETED) cmdsSucceeded++
                        else if (status == AppOperationStatus.FAILED) cmdsFailed++
                        content.v1.durationMs?.let { cmdDurationMs += it }
                        lastWasUser = false
                    }
                    is HydratedConversationItemContent.FileChange -> {
                        files += content.v1.changes.size
                        lastWasUser = false
                    }
                    is HydratedConversationItemContent.McpToolCall -> {
                        mcpCalls++
                        lastWasUser = false
                    }
                    is HydratedConversationItemContent.DynamicToolCall -> {
                        mcpCalls++
                        lastWasUser = false
                    }
                    is HydratedConversationItemContent.WebSearch -> {
                        webSearches++
                        lastWasUser = false
                    }
                    else -> {
                        lastWasUser = false
                    }
                }
            }

            return ConversationStatistics(
                totalMessages = userMessages + assistantMessages,
                userMessageCount = userMessages,
                assistantMessageCount = assistantMessages,
                turnCount = turns,
                commandsExecuted = cmdsExecuted,
                commandsSucceeded = cmdsSucceeded,
                commandsFailed = cmdsFailed,
                filesChanged = files,
                totalCommandDurationMs = cmdDurationMs,
                mcpToolCallCount = mcpCalls,
                webSearchCount = webSearches,
            )
        }
    }
}

data class ServerUsageData(
    val tokensByThread: List<Pair<String, Long>> = emptyList(),
    val activityByDay: List<Pair<LocalDate, Int>> = emptyList(),
    val modelUsage: List<Pair<String, Int>> = emptyList(),
    val rateLimits: uniffi.codex_mobile_client.RateLimitSnapshot? = null,
) {
    companion object {
        fun compute(
            threads: List<AppThreadSnapshot>,
            server: AppServerSnapshot,
        ): ServerUsageData {
            // Token usage by thread
            val tokensByThread = threads.mapNotNull { thread ->
                val tokens = thread.contextTokensUsed?.toLong() ?: return@mapNotNull null
                val title = thread.info.title?.takeIf { it.isNotBlank() } ?: "Untitled"
                title to tokens
            }.sortedByDescending { it.second }

            // Activity by day — use created_at/updated_at timestamps
            val activityMap = LinkedHashMap<LocalDate, Int>()
            for (thread in threads) {
                val ts = thread.info.updatedAt ?: thread.info.createdAt ?: continue
                val date = Instant.ofEpochSecond(ts).atZone(ZoneId.systemDefault()).toLocalDate()
                activityMap[date] = (activityMap[date] ?: 0) + 1
            }
            val activityByDay = activityMap.entries
                .map { it.key to it.value }
                .sortedBy { it.first }

            // Model usage breakdown
            val modelMap = LinkedHashMap<String, Int>()
            for (thread in threads) {
                val model = thread.model ?: thread.info.model ?: continue
                modelMap[model] = (modelMap[model] ?: 0) + 1
            }
            val modelUsage = modelMap.entries
                .map { it.key to it.value }
                .sortedByDescending { it.second }

            return ServerUsageData(
                tokensByThread = tokensByThread,
                activityByDay = activityByDay,
                modelUsage = modelUsage,
                rateLimits = server.rateLimits,
            )
        }
    }
}
